// @hyperframes/core/dist/compiler/htmlDocument.js
import { parseHTML } from "linkedom";
export const RUNTIME_BOOTSTRAP_ATTR = "data-hyperframes-preview-runtime";
const RUNTIME_SRC_MARKERS = [
    "hyperframe.runtime.iife.js",
    "hyperframes-runtime.modular.inline.js",
    "hyperframe-runtime.modular-runtime.inline.js",
    RUNTIME_BOOTSTRAP_ATTR,
];
const RUNTIME_INLINE_MARKERS = [
    "__hyperframeRuntimeBootstrapped",
    "__hyperframeRuntime",
    "__hyperframeRuntimeTeardown",
    "window.__player =",
    "window.__playerReady",
    "window.__renderReady",
];
/**
 * Parse a full HTML document or wrap a fragment so linkedom consistently puts
 * fragment content under document.body.
 */
export function parseHTMLContent(html) {
    const trimmed = html.trimStart().toLowerCase();
    if (trimmed.startsWith("<!doctype") || trimmed.startsWith("<html")) {
        return parseHTML(html).document;
    }
    return parseHTML(`<!DOCTYPE html><html><head></head><body>${html}</body></html>`).document;
}
export function stripEmbeddedRuntimeScripts(html) {
    if (!html)
        return html;
    const loweredHtml = html.toLowerCase();
    let output = "";
    let cursor = 0;
    while (cursor < html.length) {
        const scriptStart = findScriptStart(loweredHtml, cursor);
        if (scriptStart === -1) {
            output += html.slice(cursor);
            break;
        }
        output += html.slice(cursor, scriptStart);
        const startTagEnd = findTagEnd(html, scriptStart + 1);
        if (startTagEnd === -1) {
            output += html.slice(scriptStart);
            break;
        }
        const closeTagEnd = findScriptCloseTagEnd(loweredHtml, startTagEnd + 1);
        const scriptEnd = closeTagEnd === -1 ? html.length : closeTagEnd;
        const block = html.slice(scriptStart, scriptEnd);
        if (!shouldStripRuntimeScriptBlock(block)) {
            output += block;
        }
        cursor = scriptEnd;
    }
    return output;
}
function findScriptStart(loweredHtml, from) {
    let index = loweredHtml.indexOf("<script", from);
    while (index !== -1) {
        const next = loweredHtml[index + "<script".length] ?? "";
        if (isTagBoundary(next))
            return index;
        index = loweredHtml.indexOf("<script", index + 1);
    }
    return -1;
}
function findTagEnd(html, from) {
    let quote;
    for (let index = from; index < html.length; index += 1) {
        const char = html[index];
        if (quote) {
            if (char === quote)
                quote = undefined;
            continue;
        }
        if (char === '"' || char === "'") {
            quote = char;
            continue;
        }
        if (char === ">")
            return index;
    }
    return -1;
}
function findScriptCloseTagEnd(loweredHtml, from) {
    let index = loweredHtml.indexOf("</script", from);
    while (index !== -1) {
        const closeTagEnd = findScriptCloseTagBoundary(loweredHtml, index + "</script".length);
        if (closeTagEnd !== -1)
            return closeTagEnd;
        index = loweredHtml.indexOf("</script", index + 1);
    }
    return -1;
}
function findScriptCloseTagBoundary(loweredHtml, from) {
    let cursor = from;
    while (cursor < loweredHtml.length && isHtmlWhitespace(loweredHtml[cursor] ?? "")) {
        cursor += 1;
    }
    return loweredHtml[cursor] === ">" ? cursor + 1 : -1;
}
function shouldStripRuntimeScriptBlock(block) {
    const lowered = block.toLowerCase();
    for (const marker of RUNTIME_SRC_MARKERS) {
        if (lowered.includes(marker.toLowerCase()))
            return true;
    }
    for (const marker of RUNTIME_INLINE_MARKERS) {
        if (block.includes(marker))
            return true;
    }
    return false;
}
function isTagBoundary(char) {
    return char === "" || char === ">" || char === "/" || isHtmlWhitespace(char);
}
function isHtmlWhitespace(char) {
    return char === " " || char === "\n" || char === "\t" || char === "\r" || char === "\f";
}
function escapeInlineScriptSource(source) {
    return escapeCaseInsensitiveToken(escapeCaseInsensitiveToken(source, "</script", "<\\/script"), "<!--", "<\\!--");
}
function escapeCaseInsensitiveToken(source, token, replacement) {
    const loweredSource = source.toLowerCase();
    const loweredToken = token.toLowerCase();
    let output = "";
    let cursor = 0;
    while (cursor < source.length) {
        const tokenStart = loweredSource.indexOf(loweredToken, cursor);
        if (tokenStart === -1) {
            output += source.slice(cursor);
            break;
        }
        output += source.slice(cursor, tokenStart) + replacement;
        cursor = tokenStart + token.length;
    }
    return output;
}
function inlineScriptTags(scripts) {
    return scripts.map((source) => `<script>${escapeInlineScriptSource(source)}</script>`).join("\n");
}
export function injectScriptsAtHeadStart(html, scripts) {
    if (scripts.length === 0)
        return html;
    const headTags = inlineScriptTags(scripts);
    if (html.includes("<head")) {
        return html.replace(/<head\b[^>]*>/i, (match) => `${match}\n${headTags}`);
    }
    if (html.includes("<body")) {
        return html.replace("<body", () => `${headTags}\n<body`);
    }
    return `${headTags}\n${html}`;
}
export function injectScriptsIntoHtml(html, headScripts, bodyScripts, stripEmbeddedRuntime = true) {
    if (stripEmbeddedRuntime) {
        html = stripEmbeddedRuntimeScripts(html);
    }
    if (headScripts.length > 0) {
        const headTags = inlineScriptTags(headScripts);
        if (html.includes("</head>")) {
            // Function replacement avoids `$&` interpolation in runtime source.
            html = html.replace("</head>", () => `${headTags}\n</head>`);
        }
        else if (html.includes("<body")) {
            html = html.replace("<body", () => `${headTags}\n<body`);
        }
        else {
            html = `${headTags}\n${html}`;
        }
    }
    if (bodyScripts.length > 0) {
        const bodyTags = inlineScriptTags(bodyScripts);
        if (html.includes("</body>")) {
            html = html.replace("</body>", () => `${bodyTags}\n</body>`);
        }
        else {
            html = `${html}\n${bodyTags}`;
        }
    }
    return html;
}
