// @hyperframes/core/dist/core.types.js
// ── Shared cross-package types ──────────────────────────────────────────────
export function toFps(input) {
    if (typeof input === "number") {
        return { num: input, den: 1 };
    }
    return input;
}
/**
 * Decimal value of an {@link Fps} rational. Used at sites that need a
 * `number` for arithmetic (frame-index → time, frame intervals, telemetry
 * payloads) where the small precision loss of the decimal is acceptable.
 */
export function fpsToNumber(fps) {
    return fps.num / fps.den;
}
/**
 * FFmpeg-style fps argument. Returns `"30"` for integer fps and `"30000/1001"`
 * for rationals — both forms are accepted verbatim by FFmpeg's `-r` and
 * `-framerate` flags. We keep integer fps as a bare integer so existing
 * snapshot tests / log output don't churn for the common case.
 */
export function fpsToFfmpegArg(fps) {
    return fps.den === 1 ? String(fps.num) : `${fps.num}/${fps.den}`;
}
/**
 * Parse a user-supplied fps spec into an {@link Fps} rational.
 *
 * Accepted forms:
 * - integer string `"30"` → `{ num: 30, den: 1 }`
 * - integer number `30` → `{ num: 30, den: 1 }`
 * - rational string `"30000/1001"` → `{ num: 30000, den: 1001 }` (exact NTSC)
 *
 * Rejected:
 * - empty / non-numeric input
 * - decimals like `"29.97"` — callers must spell rationals with `/` so the
 *   exact denominator is unambiguous (FFmpeg treats `29.97` as a slightly
 *   different framerate than `30000/1001`).
 * - division by zero, negative or zero numerator
 * - decimal value outside `[1, 240]` — defensive bounds for "human" fps
 *   ranges (24, 25, 30, 50, 60, 120, 240, plus the NTSC trio).
 */
export function parseFps(input) {
    if (typeof input === "number") {
        if (!Number.isFinite(input))
            return { ok: false, reason: "not-a-number" };
        if (!Number.isInteger(input))
            return { ok: false, reason: "ambiguous-decimal" };
        if (input <= 0)
            return { ok: false, reason: "non-positive" };
        if (input > 240)
            return { ok: false, reason: "out-of-range" };
        return { ok: true, value: { num: input, den: 1 } };
    }
    const raw = input.trim();
    if (raw === "")
        return { ok: false, reason: "empty" };
    if (raw.includes("/")) {
        const parts = raw.split("/");
        if (parts.length !== 2)
            return { ok: false, reason: "invalid-fraction" };
        const num = Number(parts[0]);
        const den = Number(parts[1]);
        if (!Number.isFinite(num) || !Number.isFinite(den)) {
            return { ok: false, reason: "not-a-number" };
        }
        if (!Number.isInteger(num) || !Number.isInteger(den)) {
            return { ok: false, reason: "invalid-fraction" };
        }
        if (den <= 0)
            return { ok: false, reason: "invalid-fraction" };
        if (num <= 0)
            return { ok: false, reason: "non-positive" };
        const decimal = num / den;
        if (decimal < 1 || decimal > 240)
            return { ok: false, reason: "out-of-range" };
        return { ok: true, value: { num, den } };
    }
    // Integer-only path — reject `"29.97"` so users are explicit about the
    // exact rational they want.
    if (!/^-?\d+$/.test(raw)) {
        // Allow caller to differentiate "29.97" from "abc" if they want; both
        // are user errors but the message can be friendlier for decimals.
        if (/^-?\d*\.\d+$/.test(raw))
            return { ok: false, reason: "ambiguous-decimal" };
        return { ok: false, reason: "not-a-number" };
    }
    const n = Number(raw);
    if (n <= 0)
        return { ok: false, reason: "non-positive" };
    if (n > 240)
        return { ok: false, reason: "out-of-range" };
    return { ok: true, value: { num: n, den: 1 } };
}
/**
 * Convenience wrapper around {@link parseFps} for callsites that want the
 * default-30-fps fallback when input is `undefined`. Does NOT swallow parse
 * errors — those still surface via the discriminated result.
 */
export function parseFpsWithDefault(input) {
    if (input === undefined || input === "")
        return { ok: true, value: { num: 30, den: 1 } };
    return parseFps(input);
}
export const CANVAS_DIMENSIONS = {
    landscape: { width: 1920, height: 1080 },
    portrait: { width: 1080, height: 1920 },
    "landscape-4k": { width: 3840, height: 2160 },
    "portrait-4k": { width: 2160, height: 3840 },
    square: { width: 1080, height: 1080 },
    "square-4k": { width: 2160, height: 2160 },
};
// `Object.keys` ordering matches insertion order in `CANVAS_DIMENSIONS` on
// every supported JS engine; tests pin the order in `index.test.ts`. Reorder
// the table above with care.
export const VALID_CANVAS_RESOLUTIONS = Object.keys(CANVAS_DIMENSIONS);
const RESOLUTION_ALIASES = {
    "1080p": "landscape",
    hd: "landscape",
    "1080p-portrait": "portrait",
    "portrait-1080p": "portrait",
    "4k": "landscape-4k",
    uhd: "landscape-4k",
    "4k-portrait": "portrait-4k",
    "1080p-square": "square",
    "square-1080p": "square",
    "4k-square": "square-4k",
};
/**
 * Map a user-facing resolution string (canonical name or alias) to a
 * `CanvasResolution`. Returns undefined for unknown values so callers
 * can produce their own "invalid" UX (CLI exit, route validation, etc.).
 */
export function normalizeResolutionFlag(input) {
    if (!input)
        return undefined;
    const lowered = input.toLowerCase();
    if (VALID_CANVAS_RESOLUTIONS.includes(lowered)) {
        return lowered;
    }
    return RESOLUTION_ALIASES[lowered];
}
/**
 * Runtime list of every valid `CompositionVariableType`. Use this anywhere
 * a Set/array of valid type strings is needed (lint rules, validators).
 * The `satisfies` guard turns adding a new variant to the union without
 * also adding it here into a compile error.
 */
export const COMPOSITION_VARIABLE_TYPES = [
    "string",
    "number",
    "color",
    "boolean",
    "enum",
];
export function isStringVariable(v) {
    return v.type === "string";
}
export function isNumberVariable(v) {
    return v.type === "number";
}
export function isColorVariable(v) {
    return v.type === "color";
}
export function isBooleanVariable(v) {
    return v.type === "boolean";
}
export function isEnumVariable(v) {
    return v.type === "enum";
}
export function isTextElement(el) {
    return el.type === "text";
}
export function isMediaElement(el) {
    return el.type === "video" || el.type === "image" || el.type === "audio";
}
export function isCompositionElement(el) {
    return el.type === "composition";
}
export const TIMELINE_COLORS = {
    video: "#ec4899",
    image: "#3b82f6",
    text: "#06b6d4",
    audio: "#10b981",
    composition: "#f97316",
};
export const DEFAULT_DURATIONS = {
    video: 5,
    image: 5,
    text: 2,
    audio: 5,
    composition: 5,
};
export function getDefaultStageZoom(resolution) {
    const { width, height } = CANVAS_DIMENSIONS[resolution];
    return {
        scale: 1,
        focusX: width / 2,
        focusY: height / 2,
    };
}
